src/main/perl/FileWriter.pm

```diff
@@ -179,43 +180,51 @@ secure way (not following symlinks, etc).
 
 Under a verbose level, it will show in the standard output a diff of
 the old and the newly-generated contents for this file before actually
-saving to disk. This diff will B<not> be stored in any logs to prevent
-any leakages of confidential information (f.i. when writing to
-/etc/shadow).
+saving to disk.
 
 =cut
 
 sub close
 {
     my $self = shift;
 
-    my ($str, $changed, $cmd, $diff);
+    my ($changed, $diff);
     my $filename = *$self->{filename};
 
     if (*$self->{save}) {
-        # We have to do this because Text::Diff is not present in SL5. :(
-        if ($self->is_verbose() && -e $filename && *$self->{buf}) {
-            $cmd = CAF::Process->new (["diff", "-u", $filename, "-"],
-                                      stdin => "$self", stdout => \$diff,
-                                      keeps_state => 1);
-            $cmd->execute();
-            if ($diff) {
+        *$self->{save} = 0;
+        my $content_ref = *$self->{buf};
+
+        # Always read and (try to) determine the diff
+        $self->debug(2, "Reading initial contents from $filename");
+        my $content_orig = LC::File::file_contents($filename);
+
+        $diff = diff(\$content_orig, $content_ref, { STYLE => "Unified" });
+        $changed = $diff ? 1 : 0;
+
+        my $msg = 'was';
+
+        if ($changed) {
+            if($self->is_verbose()) {
                 $self->verbose ("Changes to $filename:");
                 $self->report ($diff);
+            }
+
+            if (*$self->{options}->{noaction}) {
```

Comment:
Is this test the right way round? i.e. this seems to be the block that's executed when noaction is true?